src/ML/src/MANNAutoregressive.cpp

```diff
@@ -0,0 +1,706 @@
+                                  const Contacts::EstimatedContact& newContact,
+                                  Contacts::EstimatedContact& out) -> bool {
+        if (out.index != newContact.index)
+        {
+            log()->error("{} The index of the left {} is different from the one provided. "
+                         "Provided: {}, Expected: {}.",
+                         logPrefix,
+                         footName,
+                         newContact.index,
+                         out.index);
+            return false;
+        }
+        out = newContact;
+        return true;
+    };
+
+    if (!updateFoot("left", leftFoot, m_pimpl->output.leftFoot)
+        || !updateFoot("right", rightFoot, m_pimpl->output.rightFoot))
+    {
+        return false;
+    }
+
+    m_pimpl->mannInput = input;
+    m_pimpl->currentTime = time;
+    m_pimpl->isRobotStopped = true;
+
+    if (!m_pimpl->kinDyn.setRobotState(basePosition.transform(),
+                                       input.jointPositions,
+                                       baseVelocity.coeffs(),
+                                       input.jointVelocities,
+                                       m_pimpl->gravity))
+    {
+        log()->error("{} Unable to reset the kindyncomputations object.", logPrefix);
+        return false;
+    }
+
+    // the output is not valid since we are resetting the trajectory planner
+    m_pimpl->isOutputValid = false;
+    m_pimpl->supportCornerIndex = -1;
+    m_pimpl->supportFootPtr = nullptr;
+    m_pimpl->state = autoregressiveState;
+
+    if (!m_pimpl->baseOrientationDynamics->setState({basePosition.quat()}))
+    {
+        log()->error("{} Unable to reset the base orientation dynamics.", logPrefix);
+        return false;
+    }
+
+    if (!m_pimpl->mann.setInput(m_pimpl->mannInput))
+    {
+        log()->error("{} Unable to set the inputs to MANN.", logPrefix);
+        return false;
+    }
+
+    m_pimpl->fsmState = Impl::FSM::Reset;
+
+    return true;
+}
+
+bool MANNAutoregressive::setInput(const Input& input)
+{
+    // If the output of the network is valid, then the network has been called at least once
+    constexpr auto logPrefix = "[MANNAutoregressive::setInput]";
+
+    if (m_pimpl->fsmState != Impl::FSM::Reset && !this->isOutputValid())
+    {
+        log()->error("{} The network has not been reset. Please call reset().", logPrefix);
+        return false;
+    }
+
+    if (!this->isOutputValid())
+    {
+        return true;
+    }
+
+    // get the output
+    const auto& mannOutput = m_pimpl->mann.getOutput();
+
+    // the joint positions and velocities are considered as new input
+    m_pimpl->mannInput.jointPositions = m_pimpl->mann.getOutput().jointPositions;
+    m_pimpl->mannInput.jointVelocities = m_pimpl->mann.getOutput().jointVelocities;
+
+    // TODO here we assume that the x direction of the chest frame points forward. This may
+    // cause issues on some robots (e.g., iCubV2.5)
```

Comment:
One suggestion could be to get the forward direction from config file (maybe passing directly a char 'x', 'y' or 'z')

Reply:
Yes I agree, however, since we are really in the early stage we can proceed in this way:
1. expose the `forward)_direction` in the configuration file. the parameter will be a string contacting 'x', 'y' or 'z', and for the time being we support only 'x' .(i.e., it will return an error if the parameter is different from 'x'. So at least we avoid to give the user the possibility to test the network on robots in which the x-direction points forward

Reply:
Sounds good!